### DISCUSSION TITLE
How to make `exportToSvg` embed scene content ?

### DISCUSSION THREAD

**Discussion by bric3** [excalidraw/excalidraw#3756] [Q&A] [answered]
Hi everyone,

I come from this issue https://github.com/excalidraw/excalidraw/issues/3737, and specifically this comment https://github.com/excalidraw/excalidraw/issues/3737#issuecomment-864622793, as advised by @ad1992.

Here's the comment, verbatim.

----

@ad1992 Hello again, I have a question again on this scene embedding. I'd like to do that programmatically. However I can't seem to get excalidraw embed the scene when calling `exportToSvg`

```
import React from "react";
import Excalidraw, {
    exportToBlob,
    exportToSvg,
    getSceneVersion,
    serializeAsJSON,
} from "@excalidraw/excalidraw";

// placeholder for functions
window.saveAsSvg = null;


...

export default function App() {
    const excalidrawRef = React.useRef(null);

    ...

    window.saveAsSvg = (exportParams) => {
        return exportToSvg({
            elements: excalidrawRef.current.getSceneElements(),
            appState: {
                ...excalidrawRef.current.getAppState(),
                ...exportParams
            },
        });
    };

    return (
        <div className="excalidraw-wrapper">
            <Excalidraw
                ref={excalidrawRef}
                initialData={{ appState: {
                        exportEmbedScene: true
                    }
                }}
                ...
            />
        </div>
    );
}
```

I also tried to pass the `exportEmbedScene` to the `exportToSvg` directly (as part of the `appState`). But this doesn't seem to work. The scene never gets embedded. Looking at the code 

https://github.com/excalidraw/excalidraw/blob/5cd921549a7e2b67219ee3f10d98228e23103c0f/src/scene/export.ts#L68-L134

I believe `exportToSvg` cannot embed the scene. It's likely the same for the png / jpg exports. Is it expected ?

**Comment by ad1992** [comment, marked as answer]
@bric3 thanks for pointing this out, yes you are right now the utils which are exported `exportToSvg`, `exportToPng` do not support embedding the scene.
If you notice it expects a [`metadata`](https://github.com/excalidraw/excalidraw/blob/969d3c694aedba0fdf74f8a7f727118655acba84/src/packages/utils.ts#L81) attribute, whatever is send in `metadata` will be embedded in the SVG/PNG.
The way it works in `excalidraw` is we send the `metadata` as shown [here](https://github.com/excalidraw/excalidraw/blob/master/src/data/index.ts#L44).  We will add this support soon but till then as a workaround, if you are blocked at this you will have to pass `metadata` so that scene is embedded.
You can clone [encodeSvgMetadata](https://github.com/excalidraw/excalidraw/blob/969d3c694aedba0fdf74f8a7f727118655acba84/src/data/image.ts#L91) and [encodePngMetadata](https://github.com/excalidraw/excalidraw/blob/969d3c694aedba0fdf74f8a7f727118655acba84/src/data/image.ts#L39) to get the metadata and pass it to the export utilites.

**Reply by ad1992** [reply]
@bric3 create the issue for the same referencing this QnA, you can track https://github.com/excalidraw/excalidraw/issues/3758 for the updates.

**Reply by bric3** [reply]
Thanks, I'll keep an eye on #3758 !

**Comment by bric3** [comment]
@ad1992 Hi !

I noticed quite a few image types in a recent version of Excalidraw (16), I know there's 17 already.

```typescript
export declare const IMAGE_MIME_TYPES: {
    readonly svg: "image/svg+xml";
    readonly png: "image/png";
    readonly jpg: "image/jpeg";
    readonly gif: "image/gif";
    readonly webp: "image/webp";
    readonly bmp: "image/bmp";
    readonly ico: "image/x-icon";
    readonly avif: "image/avif";
    readonly jfif: "image/jfif";
};
```

But I was wondering if embedding the scene was supported in some of those types (apart from PNG and SVG of course). I have doubt on the technical feasibility or practicality for bmp, gifs, icons, but for the others the question remains.

If I'm asking that, is that, if it is supported, or planed to be supported, I would like a documentation of how the scene is embedded. Because I need a way to detect the presence of scene in those binary file in another runtime (JVM based as it's for IntelliJ).

**Reply by bric3** [reply]
Actually I think I misread the code, the export clearly shows only SVG and PNG support the feature, and the enum is about image that can be embedded.

https://github.com/excalidraw/excalidraw/blob/65047cc2cb59843a4305f9088d886edd9b933585/packages/utils/export.ts#L136-L153

That said, it might be interesting to know if embedding the scene in other file format is on the radar.
